File: OJ/Leetcode/Algorithm/17. Letter Combinations of a Phone Number.py
class Solution(object):
    def letterCombinations(self, digits):
        """
        :type digits: str
        :rtype: List[str]
        """
        if not digits:
            return []
        table = ['','','abc','def','ghi','jkl','mno','pqrs','tuv','wxyz']
        queue = ['']
        for x in digits:
            new_q = []
            for y in queue:
                for a in table[int(x)]:
                    new_q.append(y+a)
            queue = new_q
        return queue

File: OJ/Leetcode/Algorithm/17. Letter Combinations of a Phone Number.cpp
class Solution {
public:
    vector<string> letterCombinations(string digits)
    {
        vector<string> result;
        if(digits.empty()) return vector<string>();
        static const vector<string> v = {"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
        result.push_back("");
        for(int i = 0 ; i < digits.size(); i++)
        {
            int num = digits[i]-'0';
            if(num < 0 || num > 9) break;
            const string& candidate = v[num];
            if(candidate.empty()) continue;
            vector<string> tmp;
            for(int j = 0 ; j < candidate.size() ; j++)
                for(int k = 0 ; k < result.size() ; k++)
                    tmp.push_back(result[k] + candidate[j]);
            result.swap(tmp);
        }
        return result;
    }
};